File: modules/dp/src/main/scala/gcp4zio/dp/DPClusterImpl.scala
package gcp4zio
package dp

import com.google.cloud.dataproc.v1._
import com.google.protobuf.Duration
import zio.{Task, ZIO}
import scala.jdk.CollectionConverters._

case class DPClusterImpl(client: ClusterControllerClient, project: String, region: String) extends DPCluster {

  /** This function creates Dataproc Software Config - Image Version, Single Node
    *
    * @param props
    *   ClusterProps
    * @return
    */
  private def createSoftwareConfig(props: ClusterProps): SoftwareConfig =
    if (props.singleNode) {
      SoftwareConfig.newBuilder().setImageVersion(props.imageVersion).putProperties("dataproc:dataproc.allow.zero.workers", "true").build()
    } else {
      SoftwareConfig.newBuilder().setImageVersion(props.imageVersion).build()
    }

  /** This function creates GCE Cluster Config - GCE Network, GCE Service Accounts, GCE Scopes
    *
    * @param props
    *   ClusterProps
    * @return
    */
  private def createGCEClusterConfig(props: ClusterProps): GceClusterConfig = {
    val gceClusterBuilder: GceClusterConfig.Builder = props.subnetUri match {
      case Some(value) =>
        GceClusterConfig
          .newBuilder()
          .setInternalIpOnly(props.internalIpOnly)
          .setSubnetworkUri(value)
          .addAllTags(props.networkTags.asJava)
      case None =>
        GceClusterConfig.newBuilder().setInternalIpOnly(props.internalIpOnly).addAllTags(props.networkTags.asJava)
    }

    props.serviceAccount match {
      case Some(value) => gceClusterBuilder.setServiceAccount(value).build()
      case _           => gceClusterBuilder.addServiceAccountScopes("https://www.googleapis.com/auth/cloud-platform").build()
    }
  }

  /** This function creates GCE Instance Config - Instance Disk, Machine Type, Number of Nodes
    *
    * @param props
    *   ClusterProps
    * @return
    */
  private def createGCEInstanceConfig(props: InstanceProps): InstanceGroupConfig = {
    val diskConfigM = DiskConfig
      .newBuilder()
      .setBootDiskType(props.bootDiskType)
      .setBootDiskSizeGb(props.bootDiskSizeGb)
      .build()
    InstanceGroupConfig
      .newBuilder()
      .setMachineTypeUri(props.machineType)
      .setNumInstances(props.numInstance)
      .setDiskConfig(diskConfigM)
      .build()
  }

  /** This function creates Dataproc Cluster Config combining all configs
    *
    * @param props
    *   ClusterProps
    * @return
    */
  private def createClusterConfig(props: ClusterProps): ClusterConfig = {
    if (props.singleNode)
      logger.info(s"Creating single node cluster creation request")
    else
      logger.info(s"Creating multi node cluster creation request")

    val softwareConfig = createSoftwareConfig(props)

    val gceClusterConfig = createGCEClusterConfig(props)

    val masterConfig = createGCEInstanceConfig(props.masterInstanceProps)

    val endPointConfig = EndpointConfig.newBuilder().setEnableHttpPortAccess(true).build()

    val clusterConfigBuilder =
      if (props.singleNode)
        ClusterConfig.newBuilder
          .setMasterConfig(masterConfig)
          .setSoftwareConfig(softwareConfig)
          .setConfigBucket(props.bucketName)
          .setGceClusterConfig(gceClusterConfig)
          .setEndpointConfig(endPointConfig)
      else {
        val workerConfig = createGCEInstanceConfig(props.workerInstanceProps)
        ClusterConfig.newBuilder
          .setMasterConfig(masterConfig)
          .setWorkerConfig(workerConfig)
          .setSoftwareConfig(softwareConfig)
          .setConfigBucket(props.bucketName)
          .setGceClusterConfig(gceClusterConfig)
          .setEndpointConfig(endPointConfig)
      }

    props.idleDeletionDurationSecs match {
      case Some(value) =>
        clusterConfigBuilder
          .setLifecycleConfig(LifecycleConfig.newBuilder().setIdleDeleteTtl(Duration.newBuilder().setSeconds(value)))
          .build()
      case _ => clusterConfigBuilder.build()
    }
  }

  def createDataproc(clusterName: String, props: ClusterProps): Task[Cluster] = ZIO
    .fromFutureJava {
      val clusterConfig = createClusterConfig(props)

      val cluster = Cluster.newBuilder.setClusterName(clusterName).setConfig(clusterConfig).build

      val createClusterAsyncRequest = client.createClusterAsync(project, region, cluster)

      logger.info(s"Submitting cluster creation request for $clusterName")
      createClusterAsyncRequest
    }
    .tapBoth(
      e => ZIO.succeed(logger.error(s"Cluster creation failed with error ${e.getMessage}")),
      res => ZIO.succeed(logger.info(s"Cluster ${res.getClusterName} created successfully"))
    )

  def deleteDataproc(clusterName: String): Task[Unit] = ZIO
    .fromFutureJava {
      logger.info(s"Submitting cluster deletion request for $clusterName")
      client.deleteClusterAsync(project, region, clusterName)
    }
    .tapBoth(
      e => ZIO.succeed(logger.error(s"Cluster deletion failed with error ${e.getMessage}")),
      _ => ZIO.succeed(logger.info(s"Cluster $clusterName deleted successfully"))
    )
    .unit
}

File: modules/dp/src/main/java/gcp4zio/dp/DPClusterImpl.java
package gcp4zio.dp;

import com.google.api.gax.longrunning.OperationFuture;
import com.google.cloud.dataproc.v1.*;
import com.google.protobuf.Duration;
import com.google.protobuf.Empty;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class DPClusterImpl {

  private static final Logger logger = LoggerFactory.getLogger(DPClusterImpl.class);

  private final ClusterControllerClient client;
  private final String project;
  private final String region;

  public DPClusterImpl(ClusterControllerClient client, String project, String region) {
    this.client = client;
    this.project = project;
    this.region = region;
  }

  /**
   * This function creates Dataproc Software Config - Image Version, Single Node
   *
   * @param props ClusterProps
   * @return SoftwareConfig
   */
  private SoftwareConfig createSoftwareConfig(ClusterProps props) {
    if (props.singleNode()) {
      return SoftwareConfig.newBuilder()
          .setImageVersion(props.imageVersion())
          .putProperties("dataproc:dataproc.allow.zero.workers", "true")
          .build();
    } else {
      return SoftwareConfig.newBuilder().setImageVersion(props.imageVersion()).build();
    }
  }

  /**
   * This function creates GCE Cluster Config - GCE Network, GCE Service Accounts, GCE Scopes
   *
   * @param props GCEProps
   * @return GceClusterConfig
   */
  private GceClusterConfig createGCEClusterConfig(GCEProps props) {
    GceClusterConfig.Builder gceClusterBuilder =
        GceClusterConfig.newBuilder().setInternalIpOnly(props.internalIpOnly());

    if (props.subnetUri().isPresent()) {
      gceClusterBuilder.setSubnetworkUri(props.subnetUri().get()).addAllTags(props.networkTags());
    } else {
      gceClusterBuilder.addAllTags(props.networkTags());
    }

    if (props.serviceAccount().isPresent()) {
      return gceClusterBuilder.setServiceAccount(props.serviceAccount().get()).build();
    } else {
      return gceClusterBuilder
          .addServiceAccountScopes("https://www.googleapis.com/auth/cloud-platform")
          .build();
    }
  }

  /**
   * This function creates GCE Instance Config - Instance Disk, Machine Type, Number of Nodes
   *
   * @param props ClusterProps
   * @return InstanceGroupConfig
   */
  private InstanceGroupConfig createGCEInstanceConfig(InstanceProps props) {
    DiskConfig diskConfig =
        DiskConfig.newBuilder()
            .setBootDiskType(props.bootDiskType())
            .setBootDiskSizeGb(props.bootDiskSizeGb())
            .build();
    return InstanceGroupConfig.newBuilder()
        .setMachineTypeUri(props.machineType())
        .setNumInstances(props.numInstance())
        .setDiskConfig(diskConfig)
        .build();
  }

  /**
   * This function creates Dataproc Cluster Config combining all configs
   *
   * @param props ClusterProps
   * @return ClusterConfig
   */
  private ClusterConfig createClusterConfig(ClusterProps props) {
    if (props.singleNode()) {
      logger.info("Creating single node cluster creation request");
    } else {
      logger.info("Creating multi node cluster creation request");
    }

    SoftwareConfig softwareConfig = createSoftwareConfig(props);
    GceClusterConfig gceClusterConfig = createGCEClusterConfig(props.gceClusterProps());
    InstanceGroupConfig masterConfig = createGCEInstanceConfig(props.masterInstanceProps());
    EndpointConfig endPointConfig =
        EndpointConfig.newBuilder().setEnableHttpPortAccess(true).build();

    ClusterConfig.Builder clusterConfigBuilder =
        ClusterConfig.newBuilder()
            .setMasterConfig(masterConfig)
            .setSoftwareConfig(softwareConfig)
            .setConfigBucket(props.bucketName())
            .setGceClusterConfig(gceClusterConfig)
            .setEndpointConfig(endPointConfig);

    if (!props.singleNode()) {
      InstanceGroupConfig workerConfig = createGCEInstanceConfig(props.workerInstanceProps());
      clusterConfigBuilder.setWorkerConfig(workerConfig);
    }

    if (props.idleDeletionDurationSecs().isPresent()) {
      return clusterConfigBuilder
          .setLifecycleConfig(
              LifecycleConfig.newBuilder()
                  .setIdleDeleteTtl(
                      Duration.newBuilder().setSeconds(props.idleDeletionDurationSecs().get())))
          .build();
    } else {
      return clusterConfigBuilder.build();
    }
  }

  /**
   * This public function creates Dataproc Cluster
   *
   * @param cluster String
   * @param props ClusterProps
   * @return OperationFuture<Cluster, ClusterOperationMetadata>
   */
  public OperationFuture<Cluster, ClusterOperationMetadata> createDataproc(
      String cluster, ClusterProps props) {
    ClusterConfig clusterConfig = createClusterConfig(props);

    Cluster clusterConf =
        Cluster.newBuilder().setClusterName(cluster).setConfig(clusterConfig).build();

    logger.info("Submitting cluster creation request for {}", cluster);
    return client.createClusterAsync(project, region, clusterConf);
  }

  /**
   * This public function deletes Dataproc Cluster
   *
   * @param cluster String
   * @return OperationFuture<Cluster, ClusterOperationMetadata>
   */
  public OperationFuture<Empty, ClusterOperationMetadata> deleteDataproc(String cluster) {
    logger.info("Submitting cluster deletion request for {}", cluster);
    return client.deleteClusterAsync(project, region, cluster);
  }
}
